libzim/libzim.pyx

```diff
@@ -0,0 +1,286 @@
+    
+    def get_url(self):
+        raise NotImplementedError
+
+    def get_title(self):
+        raise NotImplementedError
+    
+    def is_redirect(self):
+        raise NotImplementedError
+
+    def get_mime_type(self):
+        raise NotImplementedError
+
+    def get_filename(self):
+        raise NotImplementedError
+
+    def should_compress(self):
+        raise NotImplementedError
+
+    def should_index(self):
+        raise NotImplementedError
+
+    def redirect_url(self):
+        raise NotImplementedError
+
+    def get_data(self):
+        raise NotImplementedError
+
+    @property
+    def mimetype(self):
+        return self.c_article.getMimeType().decode('UTF-8')
+    
+    @property
+    def content(self):
+        blob = self.c_article.getData()
+        content =  blob.data()[:blob.size()]
+        try:
+            return content.decode('UTF-8')
+        except UnicodeDecodeError:
+            return content
+
+    # This changes with implementation
+    @property
+    def can_write(self):
+        raise NotImplementedError
+
+
+
+#------- ZimArticle pure virtual methods --------
+cdef public api:
+    string string_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a string"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            value = func()
+            return value.encode('UTF-8')
+    
+    Blob blob_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a Blob"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        cdef ZimBlob blob = ZimBlob(b'\x00')
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            blob = func()
+            return dereference(blob.c_blob) 
+    
+    bool bool_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a bool"""
+
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            return func()
+
+    uint64_t int_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning an int"""
+
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
```

Comment:
Maybe move all this duplicate code in another function and specialize only the actual call and the type conversion.

Reply:
Done. https://github.com/openzim/python-libzim/pull/3/commits/cc31d318f78cc6eccf3091d692d7d4330eb60988